lib: posix: define _ZEPHYR_SOURCE, _DEFAULT_SOURCE, etc

cfriedt opened the pull request:
`_POSIX_C_SOURCE=200809L` should be defined by strictly conforming POSIX applications so that relevant POSIX structures and function prototypes are made visible to the application by the C library.

Since it is desired that individual applications (libraries) should have the ability to set `_POSIX_C_SOURCE` to any arbitrary value, we require a separate, indirect, (and unfortunately non-standard) way to signal to the C library that relevant POSIX structures and function prototypes.

For the Minimal libc and Picolibc, that is `_ZEPHYR_SOURCES`.

For Newlib, that is `_DEFAULT_SOURCES`, since it does not (yet) recognize `_ZEPHYR_SOURCES`.

4 commits - please see individual commit messages for details.

Commit:
lib: posix: define _ZEPHYR_SOURCE, _DEFAULT_SOURCE, etc

_POSIX_C_SOURCE=200809L should be defined by strictly conforming
POSIX applications so that relevant POSIX structures and
function prototypes are made visible to the application by the
C library.

Since it is desired that individual applications (libraries)
should have the ability to set _POSIX_C_SOURCE to any arbitrary
value, we require a separate, indirect, (and unfortunately
non-standard) way to signal to the C library that relevant
POSIX structures and function prototypes.

For the Minimal libc and Picolibc, that is _ZEPHYR_SOURCES.

For Newlib, that is _DEFAULT_SOURCES.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
@sylvioalves - it looks as though `esp32s3_devkitm_appcpu` overflows by 128 bytes on `main` for `tests/lib/c_lib/common`. 

I don't know if a filter is really the best way to fix the test for that platform, but I could do that here if that's what you prefer.

```shell
/opt/zephyr/zephyr-sdk-0.16.4/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `noinit' will not fit in region `dram0_0_seg'
/opt/zephyr/zephyr-sdk-0.16.4/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: region `dram0_0_seg' overflowed by 128 bytes
collect2: error: ld returned 1 exit status
```

aescolar commented on lib/CMakeLists.txt:
```diff
@@ -6,8 +6,8 @@ add_compile_options($<TARGET_PROPERTY:compiler,warning_shadow_variables>)
 add_subdirectory(libc)
-add_subdirectory(posix)
 endif()
+add_subdirectory(posix)
```
There is a fundamental incompatibility between POSIX_API and using the host libC (EXTERNAL_LIBC).
Moving this out of the if make it easier for users/developers to enable that combination in the build which will lead to more confusion.
I would have this include remain conditional CONFIG_EXTERNAL_LIBC or {to COFIG_POSIX_API and make the POSIX_API kconfig depend also on !EXTERNAL_LIBC}.

cfriedt replied:
> There is a fundamental incompatibility between POSIX_API and using the host libC (EXTERNAL_LIBC).

I would disagree here.

But please, explain how you feel there is a fundamental incompatibility.

> Moving this out of the if make it easier for users/developers to enable that combination in the build which will lead to more confusion.

I also would disagree here.

But please, explain how you feel this would be confusing.

> I would have this include remain conditional CONFIG_EXTERNAL_LIBC or {to COFIG_POSIX_API and make the POSIX_API kconfig depend also on !EXTERNAL_LIBC}.

That's both unnecessary and contrary to the direction approved by the TSC / Architecture WG.

It also sounds vaguely like how CONFIG_POSIX_API used to be incompatible  with CONFIG_ARCH_POSIX, but just wrapped in different packaging.

That's fixed now though, isn't it?

At least that's what I and many others were led to believe  [here](https://github.com/zephyrproject-rtos/zephyr/issues/45100#issuecomment-1628397622).

aescolar replied:
EXTERNAL_LIBC means the host C library (normally glibc). The host C library implements the POSIX API in itself, which for the OS related logic calls into the Linux kernel.
If you build Zephyr's POSIX compatibility layer (CONFIG_POSIX_API) with it you are going to find that many of its headers conflict with the host C library, and that you are building functions and declaring variables which are already provided by the host C library. If things are built without a lot of care, the result is that most symbols (functions and global variables) come from the host C library.
For a short period of time (2017-18) the combination was allowed in Zephyr, but that resulted in too much confusion and frustration, so we stopped supporting it.

The thing with native_sim is that it allows to build the embedded code with an embedded C library. That is why you can select CONFIG_POSIX_API with it. When you select CONFIG_POSIX_API for native_sim PICOLIBC is selected by default, and things just work.

If a developer forces a build in native_sim with CONFIG_EXTERNAL_LIBC and CONFIG_POSIX_API (which by default is prevented, but can be enabled by doing things like moving this line in question out of the `if`, between other changes), that developer would be forcefully replicating the native_posix incompatibility in native_sim.

cfriedt replied:
> EXTERNAL_LIBC means the host C library (normally glibc).

It is defined as "build with an external/user-provided libc", so in the context of arch posix, that would mean the host libc (typically), but really it is any user-provided libc, which *does* include those which do not have posix support built-in.

So yes, POSIX_API is not mutually exclusive with `EXTERNAL_LIBC`.

> The host C library implements the POSIX API in itself, which for the OS related logic calls into the Linux kernel.

> If you build Zephyr's POSIX compatibility layer (CONFIG_POSIX_API) with it you are going to find that many of its headers conflict with the host C library, and that you are building functions and declaring variables which are already provided by the host C library. If things are built without a lot of care, the result is that most symbols (functions and global variables) come from the host C library.

Mostly agree.

> For a short period of time (2017-18) the combination was allowed in Zephyr, but that resulted in too much confusion and frustration, so we stopped supporting it.

Yes, that was bad, and it was good that we stopped it.

> The thing with native_sim is that it allows to build the embedded code with an embedded C library. That is why you can select CONFIG_POSIX_API with it. When you select CONFIG_POSIX_API for native_sim PICOLIBC is selected by default, and things just work.

Right - no issues there.

> If a developer forces a build in native_sim with CONFIG_EXTERNAL_LIBC and CONFIG_POSIX_API (which by default is prevented)

The above is probably not even an issue in this PR anymore, if we have concesus that _application_ conformance symbols are only ever locally defined for the library, which you suggested.

So I'm hoping you are still agreeable to that, and this issue can be resolved.

But, on the topic of external libc vs posix api..

Technically, anything that happens "below the line" (i.e. as part of the implementation of native_sim) can (should?) be mostly independent of what happens "above the line" (in the Zephyr application).

The underlying (and still very much present) problem is due to C's global namespace. The names of the calls that are made by native_sim conflict with names of the calls that the zephyr application can make, and we do not want the two to mix.

If that weren't an issue, then I don't believe there would be a reason to make posix api and external libc mutually exclusive.

Would you agree?

aescolar replied:
> It is defined as "build with an external/user-provided libc" ... but really it is any user-provided libc

That option help text was overly ambitious compared with what was implemented, like it happens some times.
Keith was kind enough to add CONFIG_NATIVE_LIBC recently, so the filtering can be done by that instead of EXTERNAL_LIBC if you prefer.

> The names of the calls that are made by native_sim conflict with names of the calls that the zephyr application can make, and we do not want the two to mix.

The C library native_sim uses is *not* exposed to the embedded code. That is the whole point. But when you select CONFIG_EXTERNAL_LIBC in Zephyr you are actively choosing to build the embedded code, include the POSIX API compatibility shim code with the host libC. So you cannot expect that same code to not see the host libC, because you just chose it.

cfriedt replied:
Using `if(NOT CONFIG_NATIVE_LIBC)`

aescolar reviewed (changes requested):
If the only way to set this flags were to set them globally for the whole Zephyr, a centralized way of setting them to avoid conflicts would be an improvement.
But this is not necessary. The libraries which need it can use zephyr_library_compile_definitions() (or they can be set per file).
So, if we are to improve this, I would do so by not having any component set them in `zephyr_compile_definitions()` but instead have those files/libraries which need it use `zephyr_library_compile_definitions()`.

cfriedt commented:
* reordered commits
* moved posix things to the first commit
* separated including the posix directory to its own commit

cfriedt commented:
> If the only way to set this flags were to set them globally for the whole Zephyr, a centralized way of setting them to avoid conflicts would be an improvement.

Agreed.

> But this is not necessary. The libraries which need it can use zephyr_library_compile_definitions() (or they can be set per file).

It is necessary when local defines conflict with global defines.

> So, if we are to improve this, I would do so by not having any component set them in `zephyr_compile_definitions()` but instead have those files/libraries which need it use `zephyr_library_compile_definitions()`.

That would be one possible change, although it would likely be going in the wrong direction, so it's hard to see it as an improvement.

Can you explain why the definitions in this PR are insufficient / incorrect?

aescolar commented:
> > If the only way to set this flags were to set them globally for the whole Zephyr, a centralized way of setting them to avoid conflicts would be an improvement.
> 
> Agreed.
> 
> > But this is not necessary. The libraries which need it can use zephyr_library_compile_definitions() (or they can be set per file).
> 
> It is necessary when local defines conflict with global defines.
> 
> > So, if we are to improve this, I would do so by not having any component set them in `zephyr_compile_definitions()` but instead have those files/libraries which need it use `zephyr_library_compile_definitions()`.
> 
> That would be one possible change, although it would likely be going in the wrong direction, so it's hard to see it as an improvement.
> 
> Can you explain why the definitions in this PR are insufficient / incorrect?

The thing is, there is no actual need for anybody to define these globally.
These macros are used to have the C library headers provide or not some symbol definitions. Therefore they can be set locally by who needs them, either in cmake as per file options, or `zephyr_library_compile_definitions()`, or just as defines in source before the includes.

Background: Today in the tree we have this defined globally just because it was "good enough", and it did not cause almost any trouble (It would be annoying if somebody was trying to define it for their file or library to something else, as the compiler would warn them, and they would have needed to either remove their definition, define it to the same thing, or add a undef before their definition. Though note that as today they are defined for the highest level of support, the headers would provide whatever users needed).

If you think it is not nice to define this globally, enough like to go an change it (which I do not oppose), the solution I would go with is to have the current Zephyr libraries/files which need to set it, to set it just for their own use (that is, so it is set only locally when building that file or library).
And that would include the POSIX_API also just setting it for its own use.

If, on the other hand, we would go with a solution like the one proposed in this PR, we would be cementing the "Zephyr defines this globally" situation with more cmake and kconfig glue. So if somebody were to have a library or some other source which, validly, tries to set this macros to something else, they would still see this issue.
Telling users "you are not allowed to set this macro, but must set this kconfig option instead" is not better than telling them "you are not allowed to set this macro, and anyhow it is already set to the highest number you can possible need, so you don't need to do anything more".

sylvioalves commented:
> @sylvioalves - it looks as though `esp32s3_devkitm_appcpu` overflows by 128 bytes on `main` for `tests/lib/c_lib/common`.
> 
> I don't know if a filter is really the best way to fix the test for that platform, but I could do that here if that's what you prefer.
> 
> ```shell
> /opt/zephyr/zephyr-sdk-0.16.4/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `noinit' will not fit in region `dram0_0_seg'
> /opt/zephyr/zephyr-sdk-0.16.4/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: region `dram0_0_seg' overflowed by 128 bytes
> collect2: error: ld returned 1 exit status
> ```

@cfriedt This is a known issue. Fix: #66899
So no need to filter it.

sylvioalves commented on soc/xtensa/espressif_esp32/esp32_net/Kconfig.series:
```diff
@@ -8,5 +8,6 @@ config SOC_SERIES_ESP32_NET
 	select CLOCK_CONTROL
 	select DYNAMIC_INTERRUPTS
 	select HAS_ESPRESSIF_HAL
+	select NEED_POSIX_C_SOURCE if WIFI_ESP32
```
@cfriedt, You can remove this from here and also from `esp_wifi_drv.c`.  No need for non-standard posix functions anymore in Espressif's sources. The entry in `esp_wifi_drv.c` was outdated I thinkl.

cfriedt replied:
Fixed!

keith-packard commented on lib/posix/CMakeLists.txt:
```diff
@@ -26,12 +26,16 @@ if(CONFIG_POSIX_SIGNAL)
-  # types for us. POSIX features to be formalized as part of #51211
-  zephyr_compile_options($<$<COMPILE_LANGUAGE:C>:-D_POSIX_THREADS>)
-  zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:-D_POSIX_THREADS>)
+if(CONFIG_NEED_POSIX_C_SOURCE)
```
I feel like the intent of these defines is for each .c file which requires a particular API to add the necessary defines. This allows code using different APIs to freely intermingle within the same application -- the C library can ensure that the correct APIs all work together. Having these be global defines means the entire pool of source code is required to use the same API.

cfriedt replied:
> I feel like the intent of these defines is for each .c file which requires a particular API to add the necessary defines. This allows code using different APIs to freely intermingle within the same application

For _application_ conformance (sec 2.2.1) I will agree that we can leave those specific symbols to local definitions (via `zephyr_library_compile_definitions*()` as suggested by @aescolar) with 2 caveats:

1. if `_POSIX_C_SOURCE` is undefined (and _XOPEN_SOURCE` is also not defined) by the application or library, then a default should be declared (via C header) the latest version supported by the implementation.

2. a warning or error is generated by the cmake family of functions `zephyr_compile_definitions*()` if an attempt is made to _globally_ set `_POSIX_C_SOURCE` or other reserved application conformance options.

The former is to avoid users encountering potentially hundreds of implicit function declaration warnings / errors when they want things to Just Work. FWICT, that is not the current case with picolibc or newlib which would explain the current global. Is there a better option, @keith-packard? 

The latter would serve to short-circuit a possible conflicting definition at build time, prior to compile time. So failing slightly faster.

@aescolar, @keith-packard - does ^^ sound reasonable? If so, I would be happy to make those changes in this PR.

For _implementation_ conformance, the system is required to certain feature test macros on a global level - e.g. `_POSIX_VERSION`, as well as several others listed in section 2.1.3.1 and elsewhere.

So please do not be surprised about those, because they are in the works 😁

keith-packard replied:
If `_POSIX_C_SOURCE` is undefined in the source file, the source file should expect to see only the Zephyr C library defines.

For picolibc, that is done by setting `_ZEPHYR_SOURCE`, which is done automatically by the library when `__ZEPHYR__` is defined. This does not require any setting of `_POSIX_C_SOURCE` or `_XOPEN_SOURCE`.

For C libraries which do not have a `_ZEPHYR_SOURCE` mode, the environment may do whatever is necessary to provide the Zephyr C library interfaces by default, including setting `_POSIX_C_SOURCE`.

Ideally, we'd avoid setting either `_ZEPHYR_SOURCE` or `_POSIX_C_SOURCE` when the application has requested a strict C library API via `__STRICT_ANSI__`. However, we can't do that as the compilers set this whenever we build with `-std=cxx`. But, we could check for `_BSD_SOURCE`, `_SVID_SOURCE`, `_GNU_SOURCE` or even `_DEFAULT_SOURCE` so that code from those environments get the API they are designed against. Picolibc does this already for `_ZEPHYR_SOURCE`; Zephyr could do this internally for other C libraries which require `_POSIX_C_SOURCE`.

aescolar replied:
I agree with @keith-packard .

> a warning or error is generated by the cmake family of functions zephyr_compile_definitions*() .. to short-circuit a possible conflicting definition at build time, prior to compile time. So failing slightly faster.

That sounds too complex, for too little benefit

> on a global level - e.g. _POSIX_VERSION

On unistd.h https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/unistd.h.html , not globally

cfriedt replied:
> If `_POSIX_C_SOURCE` is undefined in the source file, the source file should expect to see only the Zephyr C library defines.

But that is the opposite of what is happening today. `_POSIX_C_SOURCE=200809` is hard-coded globally via cmake by picolibc both in-tree and in-module and it probably should not be.

https://github.com/zephyrproject-rtos/zephyr/blob/915f0fd74c6065b39ab84b2ecafce57000c822a7/lib/libc/picolibc/CMakeLists.txt#L15
https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/zephyr/zephyr.cmake#L95

In picolibc, it says it's set "by Zephyr".

https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L131

and it should be, but not globally. We also want it to be flexibly settable on a per-library basis.

`_ZEPHYR_SOURCE` currently sets `__ZEPHYR_VISIBLE` which only affects string.h functions that are CX (i.e. part of posix rather than iso c). In Zephyr, Those will be moved from minimal / common libc to posix (and can be individually selected without selecting all of POSIX) (see other PRs).

> For picolibc, that is done by setting `_ZEPHYR_SOURCE`, which is done automatically by the library when `__ZEPHYR__` is defined. This does not require any setting of `_POSIX_C_SOURCE` or `_XOPEN_SOURCE`.

I think we might want to fix picolibc slightly by adding something like `|| defined(__ZEPHYR_POSIX_SOURCE)` to the line below.

https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L143

We need something that signals to picolibc (and other libcs) that Zephyr's configuration requires `_POSIX_C_SOURCE` to be defined even if it is not explicitly defined by the application or library (like `_DEFAULT_SOURCE`).

@keith-packard, I think you might have misinterpreted what my previous comment was getting at.

> For C libraries which do not have a `_ZEPHYR_SOURCE` mode, the environment may do whatever is necessary to provide the Zephyr C library interfaces by default, including setting `_POSIX_C_SOURCE`.

I guess that brings us back to the beginning.

What about the base case?

Let's assume that we have fixed picolibc's cmake files to not define `_POSIX_C_SOURCE` globally. That's good 👍

Now assume a random user has modified `samples/hello_world`, added a call to `fnmatch()`, added something like `CONFIG_FNMATCH=y` to `prj.conf`.

That person expects `fnmatch()` to be visible by default (without explicitly setting `_POSIX_C_SOURCE`) purely based on the configuration.

1. Either define `_POSIX_C_SOURCE` globally for the application (when required, but this would again incur conflicts with other local defines)
2. Use something like what was suggested above. That is not absolutely set based on `__ZEPHYR__` but is set based on Kconfig (e.g. `__ZEPHYR_POSIX_SOURCE`)

> Ideally, we'd avoid setting either `_ZEPHYR_SOURCE` or `_POSIX_C_SOURCE` when the application has requested a strict C library API via `__STRICT_ANSI__`.

Would `__ZEPHYR_POSIX_SOURCE` be acceptable?

> However, we can't do that as the compilers set this whenever we build with `-std=cxx`. But, we could check for `_BSD_SOURCE`, `_SVID_SOURCE`, `_GNU_SOURCE` or even `_DEFAULT_SOURCE` so that code from those environments get the API they are designed against.

> Picolibc does this already for `_ZEPHYR_SOURCE`; Zephyr could do this internally for other C libraries which require `_POSIX_C_SOURCE`.

It does not seem to do that at the moment.

https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L143

I would be agreeable to adding it there though as long as `_ZEPHYR_SOURCE` was used exclusively to indicate that POSIX functionality has been selected via Kconfig, and it was not set purely based on `__ZEPHYR__`.

Does that make sense @keith-packard?

cfriedt replied:
> > a warning or error is generated by the cmake family of functions zephyr_compile_definitions*() .. to short-circuit a possible conflicting definition at build time, prior to compile time. So failing slightly faster.
> 
> That sounds too complex, for too little benefit

Agreed - I tried this last night and it was far too much effort.

cfriedt replied:
ping @keith-packard - I think the latest revs is approximately what you're hoping for with `_ZEPHYR_SOURCE`. PTAL.

There are some conflicts within picolibc itself right now, so I'll leave those to you. Would be happy to review at some point if you need me.

keith-packard replied:
> But that is the opposite of what is happening today. `_POSIX_C_SOURCE=200809` is hard-coded globally via cmake by picolibc both in-tree and in-module and it probably should not be.

Yeah, we haven't gotten around to fixing that -- it was needed before picolibc provided `_ZEPHYR_SOURCE` support.
> I think we might want to fix picolibc slightly by adding something like `|| defined(_ZEPHYR_SOURCE)` to the line below.
> 
> https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L143

No, `_ZEPHYR_SOURCE` is not a superset of `_DEFAULT_SOURCE`, which is why there's a `!defined(_ZEPHYR_SOURCE)` in that line.

> That's mainly because we need something that signals to picolibc (and other libcs) that Zephyr's configuration requires `_POSIX_C_SOURCE` to be defined even if it is not explicitly defined by the application or library (like `_DEFAULT_SOURCE`).

I don't think Zephyr's configuration should ever require `_POSIX_C_SOURCE` be implicitly defined for a C library with `_ZEPHYR_SOURCE` support. Any code requiring POSIX APIs beyond those provided by `_ZEPHYR_SOURCE` can explicitly set `_POSIX_C_SOURCE`.

> Let's assume that we have fixed picolibc's cmake files to not define `_POSIX_C_SOURCE` globally. That's good 👍

Yeah, we've always planned to do this when picolibc got `_ZEPHYR_SOURCE` support, but somehow we haven't managed yet...

> That person expects `fnmatch()` to be visible by default (without explicitly setting `_POSIX_C_SOURCE`) purely based on the configuration.

I think that's the question here -- *does* the user expect this function to be globally visible without defining `_POSIX_C_SOURCE` or not?  If you add `_POSIX_C_SOURCE` to the global defines, then you're changing the API visible to the whole system, not just the portions using this non-Zephyr API.

There is at least one case where selecting `_GNU_SOURCE` changes the API vs selecting `_POSIX_C_SOURCE` -- `strerror_r` has two separate implementations depending on which of those you select. This makes me leery of anything which globally changes the API.

> > Ideally, we'd avoid setting either `_ZEPHYR_SOURCE` or `_POSIX_C_SOURCE` when the application has requested a strict C library API via `__STRICT_ANSI__`.

Ideally that's true, but alas we cannot do that with the current GCC code -- that sets `__STRICT_ANSI__` whenever `-std=cxx` is used. Check out the comment here: https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L129C1-L129C1

> As long as it is not hard-coded based on `__ZEPHYR__`, do you agree that `_ZEPHYR_SOURCE` would be acceptable?

Are you suggesting that every Zephyr file include `#define _ZEPHYR_SOURCE` to ensure that it has access to the Zephyr C library API? Because we can't add `-D_ZEPHYR_SOURCE` to the compiler command line for the same reason we can't add `-D_POSIX_C_SOURCE=200809L` -- that breaks any per-source-file API selection.

> It does not seem to do that at the moment.
> 
> https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L143

Right, that code intentionally skips checking for `_POSIX_C_SOURCE` as the current Zephyr configuration sets that globally (see the above gap in our process). The first thing to do is remove the setting of `_POSIX_C_SOURCE` from the global Zephyr compiler options, then fix picolibc to add `_POSIX_C_SOURCE` to the list of things to check when deciding to *not* select `_ZEPHYR_SOURCE`. We'll probably want to add some Zephyr version check or other mechanism in picolibc to make things work across a transition period.

> I would be agreeable to adding it there though as long as `_ZEPHYR_SOURCE` was used exclusively to indicate that POSIX functionality has been selected via Kconfig, and it was not set purely based on `__ZEPHYR__`.

I think that conflicts with the goal of having the Zephyr C library API available without any special definitions required in each source file.

> Does that make sense @keith-packard?

It feels like we're talking past each other here...

My reading of the current Zephyr C library API doc says:

 * The Zephyr C API is provided unless a source file explicitly requests a different API. Files do not need to use `#define _ZEPHYR_SOURCE` to get this API.
 * C API requests are made by adding a `#define _XXX_SOURCE` to individual source files above all `#include` statements.
 * The Zephyr C API is a compatible subset of the `_POSIX_C_SOURCE=200809L` API, so we can use kludges for non-Zephyr-aware C libraries like passing `-D_POSIX_C_SOURCE=200809L` on the command line. This requires that these C libraries handle the other `_*_SOURCE` flags correctly even when that is set, which (at least) newlib and picolibc do.
 * No `-D_*_SOURCE` parameter should ever be passed on the compiler command line, except for the above compatibility kludges.

aescolar replied:
> Now assume a random user has modified samples/hello_world, added a call to fnmatch(), added something like CONFIG_FNMATCH=y to prj.conf.
>
> That person expects fnmatch() to be visible by default (without explicitly setting _POSIX_C_SOURCE) purely based on the configuration.

I'd say no, just like in other systems.

cfriedt replied:
> > But that is the opposite of what is happening today. `_POSIX_C_SOURCE=200809` is hard-coded globally via cmake by picolibc both in-tree and in-module and it probably should not be.
> 
> Yeah, we haven't gotten around to fixing that -- it was needed before picolibc provided `_ZEPHYR_SOURCE` support.

What is `_ZEPHYR_SOURCE` exactly? More specifically, what do you anticipate it will signify and how do you feel we should use it in Zephyr within the POSIX API? How should Arbitrary Zephyr libraries use it?

Edit: Please feel free to link to the Zephyr documentation. I still can't seem to find it.

> > I think we might want to fix picolibc slightly by adding something like `|| defined(_ZEPHYR_SOURCE)` to the line below.
> > https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L143
> 
> No, `_ZEPHYR_SOURCE` is not a superset of `_DEFAULT_SOURCE`, which is why there's a `!defined(_ZEPHYR_SOURCE)` in that line.

I wasn't implying that `_ZEPHYR_SOURCE` is a superset of `_DEFAULT_SOURCE`.

However, from your [previous comment](https://github.com/zephyrproject-rtos/zephyr/pull/66998#discussion_r1436777355), you appeared to be suggesting to use `_ZEPHYR_SOURCE` as a path forward.

Can you please clarify?

> > That's mainly because we need something that signals to picolibc (and other libcs) that Zephyr's configuration requires `_POSIX_C_SOURCE` to be defined even if it is not explicitly defined by the application or library (like `_DEFAULT_SOURCE`).
> 
> I don't think Zephyr's configuration should ever require `_POSIX_C_SOURCE` be implicitly defined for a C library with `_ZEPHYR_SOURCE` support. Any code requiring POSIX APIs beyond those provided by `_ZEPHYR_SOURCE` can explicitly set `_POSIX_C_SOURCE`.

I think I misinterpreted your previous suggestion then. So back to the beginning....

How would you suggest I define `_POSIX_C_SOURCE=200809L` (or any other version , since you also suggested that previously) if doing so is currently broken when building Zephyr against both native_sim or Picolibc?

Also, an update: `zephyr_library_compile_definitions()` does not limit defines to the local library being built, so although that would have been nice if it would have worked.

https://github.com/zephyrproject-rtos/zephyr/issues/66921
 
> Yeah, we've always planned to do this when picolibc got `_ZEPHYR_SOURCE` support, but somehow we haven't managed yet...

What do you need for `_ZEPHYR_SOURCE` support?

> > That person expects `fnmatch()` to be visible by default (without explicitly setting `_POSIX_C_SOURCE`) purely based on the configuration.
> 
> I think that's the question here -- _does_ the user expect this function to be globally visible without defining `_POSIX_C_SOURCE` or not? If you add `_POSIX_C_SOURCE` to the global defines, then you're changing the API visible to the whole system, not just the portions using this non-Zephyr API.

From my perspective, as a long time Zephyr user as well as a longer time POSIX developer, I would say yes.

If I have `CONFIG_FNMATCH=y` in my `prj.conf`, I would expect that
* the declaration `fnmatch()` is visible when I include `<fnmatch.h>`
* the library containing the compiled version of that function will be linked with my application

This is exactly the way that Zephyr behaves for other standardized libraries, such as ISO C. For example, 

If I have `CONFIG_COMMON_LIBC_MALLOC=y` in my `prj.conf`, I would expect that
* the declaration of `malloc()` is visible when I include `<stdlib.h>`
* the library containing the compiled version of that function will be linked with my application

I would imagine this is... normal?

> There is at least one case where selecting `_GNU_SOURCE` changes the API vs selecting `_POSIX_C_SOURCE` -- `strerror_r` has two separate implementations depending on which of those you select. This makes me leery of anything which globally changes the API.

Sure... but nothing would actually be changing, right? We're already defining `_POSIX_C_SOURCE` all the time, for every build anyway. Additionally, we do have an implementation of `strerror_r()` (and many other functions) in tree already. They are compliant and perfectly optimized for Zephyr, which... doesn't really do GNU anyway ¯\_(ツ)_/¯.

> > As long as it is not hard-coded based on `__ZEPHYR__`, do you agree that `_ZEPHYR_SOURCE` would be acceptable?
> 
> Are you suggesting that every Zephyr file include `#define _ZEPHYR_SOURCE` to ensure that it has access to the Zephyr C library API?

No, that's ridiculous! In spite of that, I also feel like you've said more or less the same thing w.r.t. `_POSIX_C_SOURCE`, which is also kind of ridiculous.

Can you maybe clarify what you mean?

> Because we can't add `-D_ZEPHYR_SOURCE` to the compiler command line for the same reason we can't add `-D_POSIX_C_SOURCE=200809L` -- that breaks any per-source-file API selection.

Oh, but we already have added it to the command line for literally every single build.

I'm suggesting we avoid doing either of those things.

What are you suggesting??

> > It does not seem to do that at the moment.
> > https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/newlib/libc/include/sys/features.h#L143
> 
> Right, that code intentionally skips checking for `_POSIX_C_SOURCE` as the current Zephyr configuration sets that globally (see the above gap in our process). The first thing to do is remove the setting of `_POSIX_C_SOURCE` from the global Zephyr compiler options, then fix picolibc to add `_POSIX_C_SOURCE` to the list of things to check when deciding to _not_ select `_ZEPHYR_SOURCE`. We'll probably want to add some Zephyr version check or other mechanism in picolibc to make things work across a transition period.

I'm perfectly fine just using `_DEFAULT_SOURCE` as well. I feel like been asking the same question for days.

**How would you suggest that I get Picolibc to declare POSIX headers from one library to the next?**

**Better yet, from the perspective of the application, which has some C sources, a `CMakeLists.txt` file, and a `prj.conf` file, what is the best way for that application to have visibility of Picolibc's, or Newlib's POSIX declarations?

And that, but times 10000 because that's how it needs to scale. So, every line counts.**

> > I would be agreeable to adding it there though as long as `_ZEPHYR_SOURCE` was used exclusively to indicate that POSIX functionality has been selected via Kconfig, and it was not set purely based on `__ZEPHYR__`.
> 
> I think that conflicts with the goal of having the Zephyr C library API available without any special definitions required in each source file.

I don't believe it does. In any case, I'm mostly concerned with having Zephyr's POSIX API available.

> It feels like we're talking past each other here...

Same. But please, it would good to clarify. I really was hoping to be much more productive w.r.t. #51211 over the winter break, but I seem to be hitting some serious roadblocks. So any bit of clarification would help.

> My reading of the current Zephyr C library API doc says:
> 
> * The Zephyr C API is provided unless a source file explicitly requests a different API. Files do not need to use `#define _ZEPHYR_SOURCE` to get this API.

Sorry, when you mentioned `_ZEPHYR_SOURCE` previously, that was the first time I'd ever heard of it. I certainly wouldn't want to `#define _ZEPHYR_SOURCE` anywhere, because it's not a standard and it doesn't really seem to provide much in terms of functionality, and it only does things with Picolibc at the moment..

> * C API requests are made by adding a `#define _XXX_SOURCE` to individual source files above all `#include` statements.

Hmm.... I don't think that will scale very well unfortunately. Do you have a link to docs where this is writen? I'd like to see the PR where this was added..

Was there ever a task created to enumerate all of planned `_ZEPHYR_SOURCE` features?

Literally nothing comes up for `_ZEPHYR_SOURCE` on the docs website, and I don't believe I've ever happened upon it outside of when you mentioned it recently and from the 2 places it exists in picolibc.

<img width="885" alt="Screenshot 2023-12-27 at 10 59 26 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/7e7388c3-f1a0-4030-9228-69c3d2a87f74">

> * The Zephyr C API is a compatible subset of the `_POSIX_C_SOURCE=200809L` API, so we can use kludges for non-Zephyr-aware C libraries like passing `-D_POSIX_C_SOURCE=200809L` on the command line. This requires that these C libraries handle the other `_*_SOURCE` flags correctly even when that is set, which (at least) newlib and picolibc do.

Hmm... I'm not sure it will scale very well to require that portably written libraries be modified to build against Zephyr. And also... like other C libraries, and toolchains. It's kind of like .. eliminating several user bases.

Additionally does not scale because kludge definitions do not seem to be isolated from one another, one library to the next, and well, kludges..

> * No `-D_*_SOURCE` parameter should ever be passed on the compiler command line, except for the above compatibility kludges.

Right... this is getting extremely circular (the codebase seems to have regressed already when we started doing exactly the above, and it's happening on virtually all builds).

In any case, it sounds as though you need a list of functions to make visible when `_ZEPHYR_SOURCE` is defined.

Is that correct? Should we make a ticket for that? I could do it if you like. A starter list is [here](https://pastebin.com/NYhBWqED) obtained via the command below.

```shell
grep -R "[a-z_][a-z0-9_]*[\ ]\+[a-z_][a-z0-9_]*([^)]*)" include/zephyr/posix/ | grep -v "CONFIG_" | grep -v "zsock" | awk '{split($0,a,":"); print a[2];}' | grep -v "__z" | grep -v "^#if"
```

I guess also - is the `_ZEPHYR_SOURCE` approach really the right way to go? It seems like a lot of work for... not a lot of gain. Especially when it only works for 1 C library in the world.

**Wouldn't it be far simpler to have a single place where the application's `_POSIX_SOURCE` level could be set?** E.g. a `choice` would be ... ridiculously simple to implement in Kconfig

Is it best to rely on `_DEFAULT_SOURCES` for newlib for the time being?

cfriedt replied:
> > Now assume a random user has modified samples/hello_world, added a call to fnmatch(), added something like CONFIG_FNMATCH=y to prj.conf.
> > That person expects fnmatch() to be visible by default (without explicitly setting _POSIX_C_SOURCE) purely based on the configuration.

> I'd say no, just like in other systems.

Wait .. are you saying you define `_POSIX_C_SOURCE` above all of the headers that you include even when building trivial C programs on Linux?

That seems... extremely unlikely.

So... on other systems... people *do* rely on the correct version of `_POSIX_C_SOURCE` being defined... by the configuration?

keith-packard replied:
> What is `_ZEPHYR_SOURCE` exactly? More specifically, what do you anticipate it will signify and how do you feel we should use it in Zephyr within the POSIX API? How should Arbitrary Zephyr libraries use it?

Sorry, it's an internal detail of how Picolibc exposes the Zephyr APIs, but I thought it was a reasonable parallel to the existing `_POSIX_C_SOURCE`, `_GNU_SOURCE` et al. When set, picolibc exposes precisely the current Zephyr C library API without any additions. Further, it gets set when `__ZEPHYR__` is define and none of `_GNU_SOURCE`, `_BSD_SOURCE`, `_SVID_SOURCE` or `_DEFAULT_SOURCE` are set. This means that Zephyr source files that don't explicitly select an C library API get the standard Zephyr C libary API.

We *could* go further and document `_ZEPHYR_SOURCE` somewhere, but I'm not sure it's all that useful? Anything building inside of Zephyr will get `__ZEPHYR__` defined, and Picolibc will then go set `_ZEPHYR_SOURCE` which will set `_ZEPHYR_VISIBLE`. That's the internal picolibc symbol which enables all of the Zephyr extensions to the standard C library API.

> However, from your [previous comment](https://github.com/zephyrproject-rtos/zephyr/pull/66998#discussion_r1436777355), you appeared to be suggesting to use `_ZEPHYR_SOURCE` as a path forward.
> 
> Can you please clarify?

I think we're already there -- any source file without an explicit `#define _*_SOURCE` will end up getting the Zephyr C library API.

> How would you suggest I define `_POSIX_C_SOURCE=200809L` (or any other version , since you also suggested that previously) if doing so is currently broken when building Zephyr against both native_sim or Picolibc?

I think this is the crux of the issue. You want a way to have the entire application able to access the POSIX API without anything outside of the Kconfig settings. I'd rather that the core Zephyr code build against only the Zephyr C library API and that any code requiring a complete POSIX API say so explicitly, whether in per-file `#define _POSIX_C_SOURCE 200809L` or in CMakeLists.txt files for each module using POSIX APIs.

> What do you need for `_ZEPHYR_SOURCE` support?

It's been done for months; we just need to go remove the `_POSIX_C_SOURCE` bits when we have a new-enough version of picolibc. Thinking about this now, I suspect that's the real trick -- how do we get CMake to have the definition of `_POSIX_C_SOURCE` depend upon the version of picolibc that is being used? I know how to do that with autotools or meson, but cmake doesn't seem to make that easy...

> From my perspective, as a long time Zephyr user as well as a longer time POSIX developer, I would say yes.

Hrm. Even on a Linux system, you must explicitly `#define _GNU_SOURCE` to access the glibc extensions to `fnmatch` (`FNM_LEADING_DIR`, `FNM_CASEFOLD`, and `FNM_FILE_NAME`). You're used to building on native POSIX systems where the default C library API includes all of the POSIX apis. Zephyr is not such a system; use of POSIX APIs steps outside of the defaults.

The POSIX specification (https://pubs.opengroup.org/onlinepubs/007904875/functions/xsh_chap02_02.html) is also clear that any POSIX application must make sure that `_POSIX_C_SOURCE` is defined correctly before including any header files. 

> This is exactly the way that Zephyr behaves for other standardized libraries, such as ISO C.
> 
> If I have `CONFIG_COMMON_LIBC_MALLOC=y` in my `prj.conf`, I would expect that

I have a PR to fix this (#65434) so we don't need to ask for malloc, it just gets included if used (and, more importantly, `malloc_prepare` *doesn't* get included if you don't use it). This will also fix current configuration errors when an application calls `malloc` but `CONFIG_COMMON_LIBC_MALLOC=n`. There shouldn't be any Kconfig settings that set up the default Zephyr C library API; it should always be available. 

And that's the key here -- `malloc` is part of the default Zephyr C library API while `fnmatch` is not. To use `fnmatch`, you don't need any Kconfig settings, you just need to make sure `_POSIX_C_SOURCE` is defined in whatever source file references that symbol.

> No, that's ridiculous! In spite of that, I also feel like you've said more or less the same thing w.r.t. `_POSIX_C_SOURCE`, which is also kind of ridiculous.

Right, here's where we disagree -- I think that any source code using non-Zephyr C library functions should explicitly request them by defining the required `_*_SOURCE` value. Whether that's per-file (as it probably should be for in-tree code) or per-module (convenient for large existing POSIX code bases) is purely a matter of taste.

Mostly, I think the core Zephyr code should *not* be built with `_POSIX_C_SOURCE` defined; it's not POSIX code, it uses a different C library API.

> I'm perfectly fine just using `_DEFAULT_SOURCE` as well. I feel like been asking the same question for days.

`_DEFAULT_SOURCE` is even broader than `_POSIX_C_SOURCE`; that's a glibc symbol which enables POSIX along with most of the BSD and SVID symbols. I'd say we should avoid use of this; it's not defined by either the C or POSIX standards (he says, without checking).

> Edit: Or is it that POSIX prototypes will _always_ be visible? I'm ok with that too, I just wasn't sure if that was a requirement - a lot of people would prefer if POSIX declarations were optional, but I don't personally care.

Yeah, right now we've got a bug where using picolibc enables all of the POSIX APIs. As above, I think that's actually hard to fix because I don't know how to get CMake to detect when the version of picolibc being used supports `__ZEPHYR__`. Maybe we just assert that Zephyr doesn't support Picolibc before whatever version added that?

> **Better yet, from the perspective of the application, which has some C sources, a `CMakeLists.txt` file, and a `prj.conf` file, what is the best way for that application to have visibility of Picolibc's and / or Newlib's POSIX declarations?

It should follow the POSIX standard and ensure that `_POSIX_C_SOURCE` is defined before including any header files. Sticking the definition in `CMakeLists.txt` would be sufficient for this.

> I don't believe it does. In any case, I'm mostly concerned with having Zephyr's POSIX API available.

I think we need to decide whether a global `_POSIX_C_SOURCE` definition is acceptable or not, and I suspect that's a job for the TSC -- it's a policy decision which affects how Zephyr code is built for every developer.

> Hmm.... I don't think that will scale very well unfortunately. Do you have a link to docs where this is writen? I'd like to see the PR where this was added..

Yeah, not per-file, but also not globally for the entire OS and application. The POSIX standard makes it clear what applications are supposed to do, we just need to interpret that in the context of Zephyr and decide if it's OK to build the core OS exposed to the whole POSIX API or if we want to be more explicit about where the POSIX API is available.

> Was there ever a task created to enumerate all of planned `_ZEPHYR_SOURCE` features?

Yes, there's the set of C APIs beyond the C standard which Zephyr defines (https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#id31). It's a very short list at present (just `strtok_r` and `strnlen`).

> Hmm... I'm not sure it will scale very well to require that portably written libraries be modified to build against Zephyr. And also... like other C libraries, and toolchains. It's kind of like .. eliminating several user bases.

There's not a lot we can do inside of Zephyr here, other than make sure that the main Zephyr code base hews precisely to the Zephyr C library API and that the Zephyr C library API remain compatible with a documented standard C library API (POSIX) so that we can continue to use the work-around of setting `_POSIX_C_SOURCE` globally until the C library is fixed to explicitly support the Zephyr C library API in some fashion.

> Right... this is getting extremely circular (the codebase seems to have regressed already when we started doing exactly the above, and it's happening on virtually all builds).

Lemme post a PR disabling this for Picolibc builds and see how we fare.

> In any case, it sounds as though you need a list of functions to make visible when `_ZEPHYR_SOURCE` is defined.

Already sorted -- the above coding guidelines documents what APIs are part of the Zephyr C library API.

> I guess also - is the `_ZEPHYR_SOURCE` approach really the right way to go? It seems like a lot of work for... not a lot of gain. Especially when it only works for 1 C library in the world.

I hope you understand from this post that `_ZEPHYR_SOURCE` is something internal that Picolibc uses when building Zephyr code (which always defines `__ZEPHYR__`) to express the Zephyr C library API by default.

> Is it best to rely on `_DEFAULT_SOURCES` for newlib for the time being?

Probably best to use `_POSIX_C_SOURCE` for newlib -- that is the smallest API available from newlib which includes
the whole Zephyr C library API. And it has the benefit of being part of the POSIX standard.

aescolar replied:
>> How would you suggest I define `_POSIX_C_SOURCE=200809L` (or any other version , since you also suggested that previously) if doing so is currently broken when building Zephyr against both native_sim or Picolibc?
>
> I think this is the crux of the issue. You want a way to have the entire application able to access the POSIX API without anything outside of the Kconfig settings. I'd rather that the core Zephyr code build against only the Zephyr C library API and that any code requiring a complete POSIX API say so explicitly, whether in per-file `#define _POSIX_C_SOURCE 200809L` or in CMakeLists.txt files for each module using POSIX APIs.
>
>> From my perspective, as a long time Zephyr user as well as a longer time POSIX developer, I would say yes.
>
> Hrm. Even on a Linux system, you must explicitly `#define _GNU_SOURCE` to access the glibc extensions to `fnmatch` (`FNM_LEADING_DIR`, `FNM_CASEFOLD`, and `FNM_FILE_NAME`). You're used to building on native POSIX systems where the default C library API includes all of the POSIX apis. Zephyr is not such a system; use of POSIX APIs steps outside of the defaults.
>
> The POSIX specification (https://pubs.opengroup.org/onlinepubs/007904875/functions/xsh_chap02_02.html) is also clear that any POSIX application must make sure that `_POSIX_C_SOURCE` is defined correctly before including any header files. 
>
>> This is exactly the way that Zephyr behaves for other standardized libraries, such as ISO C.
>> 
>> If I have `CONFIG_COMMON_LIBC_MALLOC=y` in my `prj.conf`, I would expect that
>
> ...And that's the key here -- `malloc` is part of the default Zephyr C library API while `fnmatch` is not...
>
>> No, that's ridiculous! In spite of that, I also feel like you've said more or less the same thing w.r.t. `_POSIX_C_SOURCE`, which is also kind of ridiculous.
>
> Right, here's where we disagree -- I think that any source code using non-Zephyr C library functions should explicitly request them by defining the required `_*_SOURCE` value. Whether that's per-file (as it probably should be for in-tree code) or per-module (convenient for large existing POSIX code bases) is purely a matter of taste.
>
> Mostly, I think the core Zephyr code should *not* be built with `_POSIX_C_SOURCE` defined; it's not POSIX code, it uses a different C library API.
>
>> **Better yet, from the perspective of the application, which has some C sources, a `CMakeLists.txt` file, and a `prj.conf` file, what is the best way for that application to have visibility of Picolibc's and / or Newlib's POSIX declarations?
>
> It should follow the POSIX standard and ensure that `_POSIX_C_SOURCE` is defined before including any header files. Sticking the definition in `CMakeLists.txt` would be sufficient for this.
>
>> Hmm.... I don't think that will scale very well unfortunately. Do you have a link to docs where this is writen? I'd like to see the PR where this was added..
>
> Yeah, not per-file, but also not globally for the entire OS and application. The POSIX standard makes it clear what applications are supposed to do, we just need to interpret that in the context of Zephyr and decide if it's OK to build the core OS exposed to the whole POSIX API or if we want to be more explicit about where the POSIX API is available.

I'm in agreement with @keith-packard in all these.

aescolar replied:
Sidetrack following @keith-packard issue:

> Yeah, right now we've got a bug where using picolibc enables all of the POSIX APIs. As above, I think that's actually hard to fix because I don't know how to get CMake to detect when the version of picolibc being used supports `__ZEPHYR__`. 
> Maybe we just assert that Zephyr doesn't support Picolibc before whatever version added that?

That would be quite normal.
For the prebuilt picolibc coming via the sdk the minimum required SDK version check may help:
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/FindHostTools.cmake#L53
For the one coming from source as a west project, the general requirement is that we only need to keep Zephyr compatible with whatever we are pointing to in the west manifest at that moment.

aescolar replied:
> Also, an update: zephyr_library_compile_definitions() does not limit defines to the local library being built, although that would have been nice if it would have worked.

It does work. See answer in the ticket.

cfriedt replied:
> > Also, an update: zephyr_library_compile_definitions() does not limit defines to the local library being built, although that would have been nice if it would have worked.
> 
> It does work. See answer in the ticket.

Have you tried creating a slightly different definition in a separate CMake file? That did not work when I tried it.

E.g. even 200809L (rather than 200809) is enough to trigger a warning that is promoted to error.

But of course, that's only on my desk 🤷🏼‍♂️ and I provided the evidence that I and my compiler saw...

aescolar replied:
> Have you tried creating a slightly different definition in a separate CMake file? That did not work when I tried it.
> 
> But of course, that's only on my desk 🤷🏼‍♂️ and I provided the evidence that I and my compiler saw...

I'm sure it works.
Please check carefully what I wrote in the PR, and in case of doubt check how CMake propagates the state internally or what ends up in the ninja rules file.

keith-packard replied:
> I'm in agreement with @keith-packard in all these.

I think we've pretty much determined that this is largely a matter of taste, which means we should probably invoke arbitration as there isn't a 'right' answer.

cfriedt replied:
I'll put this into draft mode but will do another rev at some point when I have time again.

keith-packard reviewed (changes requested):
I'm not sure this is the right direction to be taking the code.

Commit:
lib: cmake: posix and external libc are not mutually exclusive

External C libraries might support POSIX. Then again, they
might not. In either case, Zephyr's POSIX API should still
work in the same, reliable manner.

This change serves to correct a minor conflation between
CONFIG_EXTERNAL_LIBC and CONFIG_HOST_LIBC. In the latter case
the current Zephyr convention is to use the POSIX API
implementation from the host C library.

This change is required to support 3rd-party toolchains with
custom C libraries, or 3rd-party ISO C libraries that do not
implement the POSIX API. Additional changes may be required
in the future.

This change is also required to simply provide a coherent
POSIX API implementation that is revision-controlled and
tightly coupled with Zephyr's kernel, networking, and other
APIs.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
test: lib: c_lib: do not define standard functions as static

Define local implementations as weak instead of static.

This avoids the following error under certain conditions.

error: static declaration of 'isinff' follows non-static declaration
   64 | static int isinff(float x)
      |            ^~~~~~
note: previous declaration of 'isinff' with type 'int(float)'
  121 | extern int isinff (float);
      |            ^~~~~~
error: static declaration of 'isnanf' follows non-static declaration
   75 | static int isnanf(float x)
      |            ^~~~~~
note: previous declaration of 'isnanf' with type 'int(float)'
  122 | extern int isnanf (float);
      |            ^~~~~~

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
scripts: checkpatch: let POSIX API to define POSIX API things

Allow the POSIX API to define POSIX API things!

This avoids annoying errors like those below.

ERROR:API_DEFINE: do not specify a non-Zephyr API for libc
lib/posix/zephyr_source.h:19:+#define _DEFAULT_SOURCE 1

ERROR:API_DEFINE: do not specify a non-Zephyr API for libc
lib/posix/zephyr_source.h:24:+#define _POSIX_SOURCE 1

ERROR:API_DEFINE: do not specify a non-Zephyr API for libc
lib/posix/zephyr_source.h:26:+#define _POSIX_C_SOURCE 200809L

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
I guess I'll just keep forking PR's to try and fix low hanging fruit

keith-packard commented:
> I guess I'll just keep forking PR's to try and fix low hanging fruit

At least you got me to post a PR removing `_POSIX_C_SOURCE` from the compiler command line when using picolibc? I wonder how long that will take to get working ...
